// @nuxt/vite-builder/dist/_chunks/manifest.mjs
import { readFileSync } from "node:fs";
import { relative, resolve } from "pathe";
import { parseQuery, parseURL, withTrailingSlash, withoutLeadingSlash } from "ufo";
import { hash } from "ohash";
import { pathToFileURL } from "node:url";
import { mkdir, rm, writeFile } from "node:fs/promises";
import { normalizeViteManifest, precomputeDependencies } from "vue-bundle-renderer";
import escapeStringRegexp from "escape-string-regexp";
import { serialize } from "seroval";
function isVue(id, opts = {}) {
	const { search } = parseURL(decodeURIComponent(pathToFileURL(id).href));
	if (id.endsWith(".vue") && !search) return true;
	if (!search) return false;
	const query = parseQuery(search);
	if (query.nuxt_component) return false;
	if (query.macro && (search === "?macro=true" || !opts.type || opts.type.includes("script"))) return true;
	const type = "setup" in query ? "script" : query.type;
	if (!("vue" in query) || opts.type && !opts.type.includes(type)) return false;
	return true;
}
function uniq(arr) {
	return Array.from(new Set(arr));
}
const IS_CSS_RE = /\.(?:css|scss|sass|postcss|pcss|less|stylus|styl)(?:\?[^.]+)?$/;
function isCSS(file) {
	return IS_CSS_RE.test(file);
}
function hashId(id) {
	return "$id_" + hash(id);
}
function toArray(value) {
	return Array.isArray(value) ? value : [value];
}
async function writeManifest(ctx, css = []) {
	const { nuxt } = ctx;
	const devClientManifest = {
		"@vite/client": {
			isEntry: true,
			file: "@vite/client",
			css,
			module: true,
			resourceType: "script"
		},
		...nuxt.options.features.noScripts === "all" ? {} : { [ctx.entry]: {
			isEntry: true,
			file: ctx.entry,
			module: true,
			resourceType: "script"
		} }
	};
	const clientDist = resolve(nuxt.options.buildDir, "dist/client");
	const serverDist = resolve(nuxt.options.buildDir, "dist/server");
	const manifestFile = resolve(clientDist, "manifest.json");
	const clientManifest = nuxt.options.dev ? devClientManifest : JSON.parse(readFileSync(manifestFile, "utf-8"));
	const manifestEntries = Object.values(clientManifest);
	const buildAssetsDir = withTrailingSlash(withoutLeadingSlash(nuxt.options.app.buildAssetsDir));
	const BASE_RE = new RegExp(`^${escapeStringRegexp(buildAssetsDir)}`);
	for (const entry of manifestEntries) {
		entry.file &&= entry.file.replace(BASE_RE, "");
		for (const item of ["css", "assets"]) entry[item] &&= entry[item].map((i) => i.replace(BASE_RE, ""));
	}
	await mkdir(serverDist, { recursive: true });
	if (ctx.config.build?.cssCodeSplit === false) {
		for (const entry of manifestEntries) if (entry.file?.endsWith(".css")) {
			const key = relative(ctx.config.root, ctx.entry);
			clientManifest[key].css ||= [];
			clientManifest[key].css.push(entry.file);
			break;
		}
	}
	const manifest = normalizeViteManifest(clientManifest);
	await nuxt.callHook("build:manifest", manifest);
	const precomputed = precomputeDependencies(manifest);
	await writeFile(resolve(serverDist, "client.manifest.mjs"), "export default " + serialize(manifest), "utf8");
	await writeFile(resolve(serverDist, "client.precomputed.mjs"), "export default " + serialize(precomputed), "utf8");
	if (!nuxt.options.dev) await rm(manifestFile, { force: true });
}
export { toArray as a, isCSS as i, IS_CSS_RE as n, uniq as o, hashId as r, isVue as s, writeManifest as t };
